esp-hal/src/spi/master.rs

```diff
@@ -452,15 +453,35 @@ pub trait HalfDuplexReadWrite {
 }
 
 /// SPI peripheral driver
-pub struct Spi<'d, T, M> {
+pub struct Spi<'d, M, T = AnySpi> {
```

Comment:
Orthogonal to the PR, but we should add a `Async`/`Blocking` generic param, as we may add async support using the FiFo in the future.

Reply:
Separate, but similar question: Do we want FullDuplex/HalfDuplex to remain as type params, or should we split them into `Spi(Dma)` and `HalfDuplexSpi(Dma)`?

Reply:
I'd like it to go away and just become methods on the SPI. `halfduplex_write(...)` and `fullduplex_write(...)`.

Arguably all writes and reads should be half duplex really, and transfer should be the only full duplex operation. 

The reason I say this is because there's no reason you can't share bus between full and half duplex devices. You shouldn't have to recreate the driver to switch between them.

Reply:
Yeah we are leaning toward this as well.

Reply:
This will be done in an upcoming PR as it's a rather substantial change.